mesonbuild/dependencies/misc.py

```diff
@@ -559,6 +580,13 @@ def shaderc_factory(env: 'Environment',
 packages['shaderc'] = shaderc_factory
 
 
+packages['atomic'] = atomic_factory = DependencyFactory(
+    'atomic',
+    [DependencyMethods.BUILTIN, DependencyMethods.SYSTEM],
+    builtin_class=AtomicBuiltinDependency,
+    system_class=AtomicSystemDependency,
```

Comment:
This has come up in Gentoo before, which also had a test to add -latomic to LDFLAGS after running some test code to see if it's necessary for a successful link.

https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=ec90f4c3a4e0ef9841dede7b18df90e2aceafd45

We (I, actually) disabled that and replaced it with some code that says "is -latomic a valid library? Good, then use it unconditionally with `-Wl,--push-state,--as-needed,-latomic,--pop-state`". The reason was that the test often misfired because the test code could be satisfied via builtins, but more advanced use cases in the software being built required explicit -latomic.

So I wonder if maybe we really need to first check DependencyMethods.SYSTEM, on the same assumption: if -latomic is a valid external library, adding it to be cautious is assumed to never be harmful. It makes the compiler command line potentially longer and the linker has to add another library to search through for available symbols, but that's not a huge problem given the alternative.

Reply:
Sure, that sounds reasonable to me. I swapped the order so the system method is checked first.